### ISSUE TITLE
Question about MaskFormer: Why the target for dice loss and focal loss are different?

### ISSUE THREAD

**Issue by YTEP-ZHI** [open-mmlab/mmdetection#8580]
### Discussed in https://github.com/open-mmlab/mmdetection/discussions/8579

<div type='discussions-op-text'>

<sup>Originally posted by **YTEP-ZHI** August 17, 2022</sup>
@chhluo Thanks for your hard work by adding maskformer to mmdet.  I'm wondering why the target for FocalLoss is 1-mask_targets (as below),
https://github.com/open-mmlab/mmdetection/blob/3b72b12fe9b14de906d1363982b9fba05e7d47c1/mmdet/models/dense_heads/maskformer_head.py#L409-L411
while the target for DiceLoss is simply mask_targets?
https://github.com/open-mmlab/mmdetection/blob/3b72b12fe9b14de906d1363982b9fba05e7d47c1/mmdet/models/dense_heads/maskformer_head.py#L398-L400
Why doing so? Looking forward to your reply.</div>

**Comment by chhluo** [comment]
loss_mask is a focal_loss, see https://github.com/open-mmlab/mmdetection/blob/3b72b12fe9b14de906d1363982b9fba05e7d47c1/mmdet/models/losses/focal_loss.py#L229

```python
# shape (3, 1)
mask_targets = torch.tensor([[0], [1], [0]])

target = F.one_hot(1 - mask_targets, num_classes=num_classes + 1) 
# output: 
[[1, 0], 
 [0, 1], 
 [1, 0]]

target = F.one_hot(mask_targets, num_classes=num_classes + 1) 
# output:
[[0, 1], 
 [1, 0], 
 [0, 1]]
```

**Comment by YTEP-ZHI** [comment]
Thank you so much for your reply, but I still have a problem. It seems that this line would be executed only when it's CPU mode (not running on GPU), but when running on GPU, it directly uses sigmoid_focal_loss from mmcv where I can't find any F.ont_hot used.
https://github.com/open-mmlab/mmdetection/blob/3b72b12fe9b14de906d1363982b9fba05e7d47c1/mmdet/models/losses/focal_loss.py#L225-L229

**Comment by YTEP-ZHI** [comment]
@chhluo  Hi, I still couldn't find that answer (The different focal loss implementations between GPU and CPU), could you explain that a little bit? Thanks a lot.

**Comment by chhluo** [comment]
The input and output of a function should be same for cpu version and gpu version by default, except device.

You can use the following code to check that

```python
from mmdet.models.losses import FocalLoss

import torch


x = torch.tensor([[0.6], [0.8], [0.2], [0.3]])
target = torch.tensor([0, 1, 0, 1])

x_gpu = x.cuda()
target_gpu = target.cuda()

loss = FocalLoss(use_sigmoid=True)

print(loss(x, 1 - target))
print(loss(x_gpu, 1 - target_gpu))
print(loss(x_gpu, target_gpu))

# output
# tensor(0.1348)
# tensor(0.1348, device='cuda:0')
# tensor(0.1684, device='cuda:0')

```

And you can read the source cuda code of focal loss and check that, see https://github.com/open-mmlab/mmcv/blob/dfb48c87ae3156c4ee5f921edf89efc0fdc0c5bc/mmcv/ops/csrc/common/cuda/sigmoid_focal_loss_cuda_kernel.cuh#L16

**Comment by YTEP-ZHI** [comment]
Thank you so much for solving my doubts.
